tracing/tracer.py

```diff
@@ -85,65 +141,89 @@ def _evaluate_object(self, class_object: typing.Any) -> dict[str, type]:
+        fwm = FrameWithMetadata(frame)
+
+        self._advance_optimisations(fwm)
+        self._update_optimisations(fwm)
+
+        # Tracing has been toggled off for this line now, simply return
+        if any(opt.status() in Optimisation.OPTIMIZING_STATES for opt in self.optimisation_stack):
+            return self._on_trace_is_called
+
+        # Ignore out of project files
+        if not pathlib.Path(frame.f_code.co_filename).is_relative_to(self.project_dir):
+            return self._on_trace_is_called
+
         code = frame.f_code
         function_name = code.co_name
         possible_class = _get_class_in_frame(frame)
 
-        # Check we do not trace somewhere we do not belong, e.g. Python's stdlib!
-        full_path = pathlib.Path(code.co_filename)
-        if not full_path.is_relative_to(self.project_dir):
-            return self._on_trace_is_called
-
         file_name = pathlib.Path(code.co_filename).relative_to(self.project_dir)
         line_number = frame.f_lineno
 
         names2types, category = None, None
         if event == "call":
+            logger.info(f"Tracing call: {inspect.getframeinfo(frame)}")
             names2types = self._on_call(frame, arg)
             category = TraceDataCategory.FUNCTION_ARGUMENT
 
         elif event == "return":
+            logger.info(f"Tracing return: {inspect.getframeinfo(frame)}")
             names2types = self._on_return(frame, arg)
             category = TraceDataCategory.FUNCTION_RETURN
 
             # Adds tracing data of class members if the return is from a class function.
             if possible_class is not None:
                 names2types2 = self._on_class_function_return(frame)
                 category2 = TraceDataCategory.CLASS_MEMBER
-                self._update_trace_data_with(file_name, possible_class, function_name, line_number, category2, names2types2)
+
+                self._update_trace_data_with(
+                    file_name,
+                    possible_class,
+                    function_name,
+                    line_number,
+                    category2,
+                    names2types2,
+                )
 
         elif event == "line":
+            logger.info(f"Tracing line: {inspect.getframeinfo(frame)}")
             names2types = self._on_line(frame)
             category = TraceDataCategory.LOCAL_VARIABLE
 
         elif event == "exception":
+            logger.info(f"Skipping exception: {inspect.getframeinfo(frame)}")
             pass
 
-        else:
-            self.stop_trace()
-            print("The event " + str(event) + " is unknown.")
-            # Note: The value error does not stop the program for some reason.
-            raise ValueError("The event" + str(event) + " is unknown.")
+        # NOTE: If there is any error occurred in the trace function, it will be unset, just like settrace(None) is called.
+        # NOTE: therefore, throwing an exception does not work, as the trace function will simply be unset
```

Comment:
While the comment is true, the code is still neccessary to determine whether the tracer has encountered an unhandled event (= default-case in switch). Add the else and remove the #Note in the else.